File: wikiQ/questions.hs
import Data.List
--Q1
myLast :: [a]->a
myLast [x] = x
myLast (_:xs) = myLast xs

--Q2
myButLast :: [a]->a 
myButLast [x,y] = x  
myButLast (_:xs) = myButLast xs

--Q3
elementAt :: [a]->Int->a
elementAt (x:xs) i | i==1 = x 
                   | otherwise = elementAt xs (i-1)

--Q4
myLength :: [a] -> Int
myLength [x] = 1
myLength (_:xs) = 1 + myLength xs
               
--Q5
myReverse :: [a]->[a]
myReverse [x] = [x]
myReverse (x:xs) = myReverse xs++[x] 

--Q6
isPalindrome :: (Eq a) => [a] -> Bool
isPalindrome (xs) = (reverse xs) == xs

--Q7
-- Don't Care

--Q8
--compress :: (Eq a) => [a] ->[a]
--compress (x:xs) = x: rest xs x
--rest :: (Eq a) => [a] -> a -> [a]
--rest [x] y | x == y = []
--          | otherwise = [x]
--rest (x:xs) y | x == y = rest xs y
--              | otherwise = [x] ++ rest xs x
compress :: (Eq a) => [a] -> [a]
compress (xs) = map head (group xs)

File: wikiQ/questions.pl
%Q1
mylast([X], X).
mylast([_|Xs], Y)  :- mylast(Xs, Y).

%Q2
mybutlast([X,_], X).
mybutlast([_|Xs], Y) :- mybutlast(Xs, Y).

%Q3
elementat([X|_], 1, X).
elementat([_|Xs], I, Y) :- I1 is I-1, elementat(Xs,I1,Y).

%Q4
mylength([], 0).
mylength([_|Xs], Y+1) :- mylength(Xs, Y).

%Q5
myreverse([], Z, Z).
myreverse([X|Xs], Y, Z) :- myreverse(Xs, Y, [X|Z]).

%Q6
palindrome(Xs) :- myreverse(Xs, Ys, []), Xs==Ys.

